accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -0,0 +1,371 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.impl;
+
+import accord.api.Agent;
+import accord.api.DataStore;
+import accord.api.Key;
+import accord.api.ProgressLog;
+import accord.local.Command;
+import accord.local.CommandStore;
+import accord.local.CommandsForKey;
+import accord.local.Node;
+import accord.primitives.KeyRange;
+import accord.primitives.KeyRanges;
+import accord.primitives.Timestamp;
+import accord.primitives.TxnId;
+import accord.local.*;
+import org.apache.cassandra.utils.concurrent.AsyncPromise;
+import org.apache.cassandra.utils.concurrent.Future;
+import org.apache.cassandra.utils.concurrent.Promise;
+
+import java.util.Collection;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+public abstract class InMemoryCommandStore extends CommandStore
+{
+    private final NavigableMap<TxnId, Command> commands = new TreeMap<>();
+    private final NavigableMap<Key, CommandsForKey> commandsForKey = new TreeMap<>();
+
+    public static InMemoryCommandStore inMemory(CommandStore commandStore)
+    {
+        return (InMemoryCommandStore) commandStore;
+    }
+
+    public InMemoryCommandStore(int generation, int index, int numShards, Function<Timestamp, Timestamp> uniqueNow, LongSupplier currentEpoch, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, RangesForEpoch rangesForEpoch)
+    {
+        super(generation, index, numShards, uniqueNow, currentEpoch, agent, store, progressLogFactory, rangesForEpoch);
+    }
+
+    @Override
+    public Command ifPresent(TxnId txnId)
+    {
+        return commands.get(txnId);
+    }
+
+    @Override
+    public Command command(TxnId txnId)
+    {
+        return commands.computeIfAbsent(txnId, id -> new InMemoryCommand(this, id));
+    }
+
+    public boolean hasCommand(TxnId txnId)
+    {
+        return commands.containsKey(txnId);
+    }
+
+    @Override
+    public CommandsForKey commandsForKey(Key key)
+    {
+        return commandsForKey.computeIfAbsent(key, InMemoryCommandsForKey::new);
+    }
+
+    public boolean hasCommandsForKey(Key key)
+    {
+        return commandsForKey.containsKey(key);
+    }
+
+    @Override
+    public CommandsForKey maybeCommandsForKey(Key key)
+    {
+        return commandsForKey.get(key);
+    }
+
+    // TODO: command store api will need to support something like this for repair/streaming
+    public void forEpochCommands(KeyRanges ranges, long epoch, Consumer<Command> consumer)
```

Comment:
I still need to look closer at our epoch logic but if we have multiple epochs that are not full acknowledged then wouldn't forEpochCommands have the wrong min epoch? This is only used in tests and not in the C* patch... if this is only for tests then I am cool with this, but it would be good to start annotating what methods are not meant to be used in `src/main`

Same comment for forCommittedInEpoch

Reply:
Reply from Blake in JIRA

No I don't think so, these 2 commands just return witnessed/committed commands belonging to the given epoch. These (and all InMemory* implementations) are only meant to be used in tests. These methods in particular are placeholders to range movements in the library can be validated and should be superseded by the work to "officially" support bootstrap, low bounds, etc.